### PR TITLE
Improve wrapper for `mlflow.trace`

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11550?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11550/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11550
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Use `try-catch` to ensure the result of the decorated function is always returned.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by BenWilson2 — APPROVED** [review]
Thanks Haru!

**Comment by BenWilson2** [comment]
We'll need to ensure that the span event is properly closed. Within the exception catch logic, can we end the span with the Exception information attached to a failed Span event? cc @B-Step62 there is a mechanism for handling this within OTeL, correct?

**Comment by B-Step62** [comment]
> We'll need to ensure that the span event is properly closed. Within the exception catch logic, can we end the span with the Exception information attached to a failed Span event? cc @B-Step62 there is a mechanism for handling this within OTeL, correct?

```
    try:
        ...
            with trace_api.use_span(span, end_on_exit=False):
                mlflow_span = MLflowSpanWrapper(span, span_type=span_type)
                mlflow_span.set_attributes(attributes or {})
                trace_manager.add_or_update_span(mlflow_span)
                yield mlflow_span
        ...
    finally:
        mlflow_span.end()
```

The `mlflow.start_span` impl also has try-catch and the span is closed in `finally` block, so it's guaranteed to be closed anyway. The try-catch added in this PR is just for extra safety🙂

**Comment by B-Step62** [comment]
> We'll need to ensure that the span event is properly closed. Within the exception catch logic, can we end the span with the Exception information attached to a failed Span event? cc @B-Step62 there is a mechanism for handling this within OTeL, correct?

The `mlflow.start_span()` has another try-catch and the span is ended at `finally` clause, so it should be guaranteed the span to be closed.

I think what happened here in the failed test is that the exception raised from the child span is eaten by the new `try-catch`, then the function is simply retried in the `except` block without decorating, resulted in a new trace (with single span) is created. What happens here is kinda complex.

TLDR: we shouldn't wrap the entire function call with try-catch cuz it will eat the valid exception from the original function and retry.

1. `predict` is called and a parent span is created.
2. `some_operation_raise_error` is called and the child span is created.
3. `some_operation_raise_error` raises exception. It closes the both spans and a trace is **logged**.
4. The exception is caught by the new try-catch.
5. It retries the `some_operation_raise_error` in `except` block, but without creating span.
6. It raises error again, which bubbles up to the parent `predict` call.
7. The new `try-catch` catches the exception, retries `predict` call without creating a span.
8. `some_operation_raise_error` executed again (with span), creates a new trace with itself alone (**<- This is what is shown in the test finally**)
9. The `predict` function is halted with the exception raised from `some_operation_raise_error`.

**Review by B-Step62 — CHANGES_REQUESTED** [review]
Could you remove the new try-catch, and instead add a comment why we do the original way?

**Comment by harupy** [comment]
@B-Step62 Thanks for the investigation. Makes sense!

**Review by B-Step62 — APPROVED** [review]
